packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -0,0 +1,271 @@
+        this.onSave();
+        this.close();
+      });
+    } else {
+      FrontendServices.putMetadata(
+        this.namespace,
+        this.name,
+        JSON.stringify(newMetadata)
+      ).then((response: any): void => {
+        this.handleDirtyState(false);
+        this.onSave();
+        this.close();
+      });
+    }
+  }
+
+  getFormItem(searchLabel: string): FormItem {
+    return this.metadataForm.find(({ value, type, label, schemaField }) => {
+      return label == searchLabel;
+    });
+  }
+
+  handleTextInputChange(event: any, label: string): void {
+    this.handleDirtyState(true);
+    this.getFormItem(label).value = event.nativeEvent.srcElement.value;
+  }
+
+  handleDropdownChange = (label: string, value: string): void => {
+    this.handleDirtyState(true);
+    this.getFormItem(label).value.choice = value;
+    this.update();
+  };
+
+  handleDirtyState(dirty: boolean): void {
+    this.dirty = dirty;
+    if (this.dirty && !this.title.className.includes(DIRTY_CLASS)) {
+      this.title.className += ` ${DIRTY_CLASS}`;
+    } else if (!this.dirty) {
+      this.title.className = this.title.className.replace(DIRTY_CLASS, '');
+    }
+  }
+
+  onAfterShow(): void {
+    if (!this.editor) {
+      const getMimeTypeByLanguage = this.editorServices.mimeTypeService
+        .getMimeTypeByLanguage;
+      this.editor = this.editorServices.factoryService.newInlineEditor({
+        host: document.getElementById('Code:' + this.name),
+        model: new CodeEditor.Model({
+          value: this.getFormItem('Code').value,
+          mimeType: getMimeTypeByLanguage({
+            name: this.getFormItem('Language').value.choice,
+            codemirror_mode: this.getFormItem('Language').value.choice
+          })
+        })
+      });
+      this.editor.model.value.changed.connect((args: any) => {
+        this.getFormItem('Code').value = args.text;
+        this.handleDirtyState(true);
+      });
+    }
+  }
+
+  render(): React.ReactElement {
+    const inputElements = [];
+    for (const field of this.metadataForm) {
+      if (field.type == 'TextInput') {
+        inputElements.push(
+          <FormGroup
+            key={field.label}
+            label={field.label}
+            labelInfo="(required)"
+            helperText={field.description}
+          >
+            <InputGroup
+              onChange={(event: any): void => {
+                this.handleTextInputChange(event, field.label);
+              }}
+              defaultValue={field.value}
+              type="text-input"
+            />
+          </FormGroup>
+        );
+      } else if (field.type == 'DropDown') {
+        inputElements.push(
+          <DropDown
+            field={field}
+            handleDropdownChange={this.handleDropdownChange}
+          ></DropDown>
+        );
+      }
+    }
+    let headerText = `Edit "${this.getFormItem('Name').value}"`;
+    if (!this.name) {
+      headerText = 'Add new metadata';
```

Comment:
Do we have the schema corresponding to the metadata being edited at this point?  If so should we use the schema name instead of `metadata`?  E.g., `Add new code-snippet`

Reply:
I hadn't gotten to this code yet, but I can do this